conans/client/installer.py

```diff
@@ -36,76 +36,225 @@ def _init_package_info(deps_graph, paths, current_path):
 
-    def report_matches(self):
-        for pattern in self._unused_patterns:
-            self._out.error("No package matching '%s' pattern" % pattern)
+        # Build step might need DLLs, binaries as protoc to generate source files
+        # So execute imports() before build, storing the list of copied_files
+        from conans.client.importer import run_imports
+        copied_files = run_imports(self._conan_file, self.build_folder, self._out)
+
+        try:
+            # This is necessary because it is different for user projects
+            # than for packages
+            logger.debug("Call conanfile.build() with files in build folder: %s"
+                         % os.listdir(self.build_folder))
+            self._out.highlight("Calling build()")
+            with conanfile_exception_formatter(str(self._conan_file), "build"):
+                self._conan_file.build()
+
+            self._out.success("Package '%s' built" % self._conan_file.info.package_id())
+            self._out.info("Build folder %s" % self.build_folder)
+        except Exception as exc:
+            os.chdir(src_folder)
+            self._out.writeln("")
+            self._out.error("Package '%s' build failed" % self._conan_file.info.package_id())
+            self._out.warn("Build folder %s" % self.build_folder)
+            if isinstance(exc, ConanExceptionInUserConanfileMethod):
+                raise exc
+            raise ConanException(exc)
+
+        finally:
+            self._conan_file._conanfile_directory = export_folder
+            # Now remove all files that were imported with imports()
+            for f in copied_files:
+                try:
+                    if f.startswith(self.build_folder):
+                        os.remove(f)
+                except Exception:
+                    self._out.warn("Unable to remove imported file from build: %s" % f)
+
+    @property
+    def build_folder(self):
+        if not self._build_folder:
+            new_id = build_id(self._conan_file)
+            new_ref = PackageReference(self._conan_ref, new_id) if new_id else self._package_reference
+            self._build_folder = self._client_cache.build(new_ref, self._conan_file.short_paths)
+        return self._build_folder
+
+
+def _raise_package_not_found_error(conan_file, conan_ref, out):
+    settings_text = ", ".join(conan_file.info.full_settings.dumps().splitlines())
+    options_text = ", ".join(conan_file.info.full_options.dumps().splitlines())
+
+    out.warn('''Can't find a '%s' package for the specified options and settings:
+- Settings: %s
+- Options: %s
+''' % (conan_ref, settings_text, options_text))
+
+    raise ConanException('''Missing prebuilt package for '%s'
+Try to build it from sources with "--build %s"
+Or read "http://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package"
+''' % (conan_ref, conan_ref.name))
+
+
+def _handle_system_requirements(conan_file, package_reference, client_cache, out):
```

Comment:
Not very happy with this function outside the classes but is shared, it's called after the building and after the remote package retrieving.